Optimize pytorch layer_norm forward

xiaomengy opened the pull request:
Summary:
Seperate from D15194600
Optimize pytorch layer_norm op part 1:
  optimize layer_norm_forward_cpu
  import Eigen Maps for the performance of reduction

Differential Revision: D15290608

soumith reviewed (changes requested):
this PR now introduces Eigen into pytorch optimizations -- without discussion.
The previous comments of moving to native/cpu and using dispatcher are not addressed

BIT-silence commented:
Removed Eigen now. For reduce case, I found that even using Vec256 is still slower than Eigen. So maybe we can consider to discuss or figure out how to make it better.

VitalyFedyunin commented on aten/src/ATen/native/layer_norm.cpp:
```diff
@@ -0,0 +1,148 @@
+#include <ATen/NativeFunctions.h>
```
Please move LayerNormForwardCPUImpl and all CPU logic into the /cpu subfolder using DECLARE_DISPATCH logic (see CopyKernel for the reference). It will allow to utilize AVX2 instructions and other optimizations of OSS build.

BIT-silence replied:
Done

VitalyFedyunin commented on aten/src/ATen/native/layer_norm.cpp:
```diff
@@ -0,0 +1,148 @@
+      X_dims.cend(),
+      1LL,
+      std::multiplies<std::int64_t>());
+  Tensor Y = at::empty_like(X);
```
I'm quite sure you want to call at:native::empty_like here and avoid all tracing/profiling checks.

VitalyFedyunin reviewed (changes requested):
This PR requires new tests to be written for native_layer_norm

cpuhrsch commented:
@BIT-silence - Is this a copy of the C2 kernels?

VitalyFedyunin commented:
You are exposing native_layer_norm function to the user level api which seems to be requiring contiguous inputs. Please add tests to cover APi calls as well as contiguous checks. Alternatively if intend is to introduce fast kernel, implement all code as kernel and avoid creating a new native function.

Also it would be nice to see benchmarks that compares calling old and new layer_norm code.

BIT-silence commented:
> @BIT-silence - Is this a copy of the C2 kernels?

Logically it is, while C2 kernel is implemented by using Eigen lib. I tested the performance difference, for the rowwise moments part, Eigen version will be faster than this version by using compiler's auto vectorization, elementwise affine part performs the same. While I also tried to use Vec256 in rowwise moments part, it actually a little slower than the for-loop with auto vectorization.

BIT-silence commented:
Thanks for the advice. I have removed native_layer_norm function. Later I will add the backward part for layer_norm then do autograd for layer_norm instead of batch_norm.

Some benchmark result for this change.
input shape = [64, 128, 56, 56] and normalized_shape = [128, 56, 56] with elementwise_affine=True,
on devvm forward time from 350ms to 87.6ms. And with this approach, it can let hugging face BERT model forward about 12% faster.

soumith reviewed (approved):
thanks for making the changes. From my side this is good to go.

BIT-silence commented:
> This PR requires new tests to be written for native_layer_norm

Actually I have removed native_layer_norm. Currently we don't expose new functions and add a fast path for layer_norm. Is that fine?

fmassa commented:
We should ideally test that the fast path gives the same results as the slow path. But maybe this is implicitly tested in the cuda tests

BIT-silence commented:
> We should ideally test that the fast path gives the same results as the slow path. But maybe this is implicitly tested in the cuda tests

Actually not only cuda, but also jit test covers that.

BIT-silence commented:
Since I will add the grad part for this fast path and then all the layer_norm on CPU should go through this path. As the current fast path is actually tested by existing tests, I'm wondering if there is any need to add some specific test for that.

xiaomengy committed:
Optimize pytorch layer_norm forward (#20345)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/20345

Seperate from D15194600
Optimize pytorch layer_norm op part 1:
  optimize layer_norm_forward_cpu
  import Eigen Maps for the performance of reduction

Reviewed By: zheng-xq

Differential Revision: D15290608

fbshipit-source-id: d5589f67c515644403ff3ad11006ec43bab18809
